From af63a907097caed40841b2c75199cbb0d501936f Mon Sep 17 00:00:00 2001
From: Patrice Chotard <patrice.chotard@foss.st.com>
Date: Wed, 17 Jan 2024 16:45:59 +0100
Subject: [PATCH 0796/1141] mtd: hyperbus: hyperbus-stm32: Add SFDP calibration
 mode

Add SFDP calibration mode for JEDEC flashes.

Change-Id: I41376d1e3dcb69c5e2bce6af87a8f886e0a1d58b
Signed-off-by: Patrice Chotard <patrice.chotard@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/353484
Domain-Review: Christophe KERELLO <christophe.kerello@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Christophe KERELLO <christophe.kerello@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/mtd/hyperbus/hyperbus-stm32.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/mtd/hyperbus/hyperbus-stm32.c b/drivers/mtd/hyperbus/hyperbus-stm32.c
index 967147dc498e..9b88eda14495 100644
--- a/drivers/mtd/hyperbus/hyperbus-stm32.c
+++ b/drivers/mtd/hyperbus/hyperbus-stm32.c
@@ -5,6 +5,7 @@
  */
 
 #include <linux/mtd/cfi.h>
+#include <linux/mtd/sfdp.h>
 #include <memory/stm32-omi.h>
 
 #define STM32_AUTOSUSPEND_DELAY -1
@@ -119,8 +120,13 @@ static int stm32_hyperbus_check_transfert(struct stm32_omi *omi)
 	cfi.interleave = 1;
 	cfi.device_type = CFI_DEVICETYPE_X16;
 
-	ret = cfi_qry_mode_on(0, map, &cfi);
-	cfi_qry_mode_off(0, map, &cfi);
+	if (omi->jedec_flash) {
+		ret = hyperbus_sfdp_mode_on(0, map, &cfi);
+		hyperbus_sfdp_mode_off(0, map, &cfi);
+	} else {
+		ret = cfi_qry_mode_on(0, map, &cfi);
+		cfi_qry_mode_off(0, map, &cfi);
+	}
 
 	return !ret;
 }
-- 
2.39.2

